### PR TITLE
USSBookmarks migration

### PR DESCRIPTION
**Pull request by darkdh**
fix https://github.com/brave/brave-browser/issues/6437

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [ ] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

### Add (1)
1. Establish sync chain without any bookmarks between device A and device B
2. Add A1.com and A2.com on device A
3. A1.com and A2.com should appear on device B shortly

### Add with existing bookmarks (2)
1. Add A1.com and A2.com on device A before sync chain creation
2. Create sync chain on device A and connect device B
3 . A1.com and A2.com should appear on device B instantly

### Add with same order (3)
1. Establish sync chain without any bookmarks between device A and device B
2. Add A1.com and A2.com on device A
3. Add B1.com and B2.com on device B
4. Wait for A1, A2, B1, B2 appeared on both devices
5. Add A3.com to device A and B3.com to device B almost at the same time
6. both devices should have same bookmarks hierarchy

### Move
1.  Follow `Add (1)`
2. Add folder A on device A
3. Move A1.com and A2.com into folder A
4. Changes should be reflected on device B

### Modify
1.  Follow `Add (1)`
2. Edit A1.com to be A1111.com
3. Change should be reflected on device B

### Reorder (1)
1.  Follow `Add (1)`
2. Move A2.com in front of A1.com
3. Change should be reflected on device B

### Reorder (2)
1.  Follow `Add (1)`
2. Move A1.com behind A2.com
3. Change should be reflected on device B

### Delete
1. Follow `Move`
2. Delete folder A on device A
3. There will soon be no folder or bookmarks on device B

### Migration
1. Do `Add (1)` from old Brave
2. Quit from both devices and launch new Brave
3. Randomly pick `Move`, `Modify`, `Reorder` or `Delete` to test

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: patches/components-sync_bookmarks-bookmark_remote_updates_handler.cc.patch [hunk 1]
```diff
@@ -0,0 +1,27 @@
+diff --git a/components/sync_bookmarks/bookmark_remote_updates_handler.cc b/components/sync_bookmarks/bookmark_remote_updates_handler.cc
+index 8b81f03c143f4f036d56d92c62f2c15ded7ec81d..85c6bcaf6d4c14c71b7f17426f415e2d1aa32f87 100644
+--- a/components/sync_bookmarks/bookmark_remote_updates_handler.cc
++++ b/components/sync_bookmarks/bookmark_remote_updates_handler.cc
+@@ -144,7 +144,7 @@ void ApplyRemoteUpdate(
+   // Compute index information before updating the |tracker|.
+   const size_t old_index = size_t{old_parent->GetIndexOf(node)};
+   const size_t new_index =
+-      ComputeChildNodeIndex(new_parent, update_entity.unique_position, tracker);
++      ComputeChildNodeIndex(new_parent, update_entity.unique_position, tracker) BRAVE_APPLY_REMOTE_UPDATE;
```
Comment (bridiver) [inline patches/components-sync_bookmarks-bookmark_remote_updates_handler.cc.patch]: do we have to change this line? Can't we check against the value of `new_index` ? I'm also concerned that we're using a ternary operator with an int. It's better to make a int value check there

Comment (bridiver) [reply patches/components-sync_bookmarks-bookmark_remote_updates_handler.cc.patch]: actually looking at this more closely I think `BRAVE_APPLY_REMOTE_UPDATE` should just const cast `new_index` and change it

File: patches/components-sync_bookmarks-bookmark_remote_updates_handler.cc.patch [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+     LogProblematicBookmark(RemoteBookmarkUpdateError::kMissingParentNode);
+     return false;
+   }
++  BRAVE_PROCESS_CREATE_1
+   const bookmarks::BookmarkNode* bookmark_node =
+       CreateBookmarkNodeFromSpecifics(
+           update_entity.specifics.bookmark(), parent_node,
+           ComputeChildNodeIndex(parent_node, update_entity.unique_position,
+-                                bookmark_tracker_),
++                                bookmark_tracker_) BRAVE_PROCESS_CREATE_2,
```
Comment (bridiver) [inline patches/components-sync_bookmarks-bookmark_remote_updates_handler.cc.patch]: same as above

File: patches/components-sync_bookmarks-bookmark_remote_updates_handler.cc.patch (hunk 1 above)
Comment (darkdh) [reply patches/components-sync_bookmarks-bookmark_remote_updates_handler.cc.patch]: fixed in 9b84b78a4497e259b5303d3420c81be09d6f1ef9

File: patches/components-sync_bookmarks-bookmark_remote_updates_handler.cc.patch (hunk 2 above)
Comment (darkdh) [reply patches/components-sync_bookmarks-bookmark_remote_updates_handler.cc.patch]: fixed in 9b84b78a4497e259b5303d3420c81be09d6f1ef9

File: app/brave_main_delegate.cc [hunk 3]
```diff
@@ -173,7 +173,6 @@ bool BraveMainDelegate::BasicStartupComplete(int* exit_code) {
       features::kWebXr.name,
       features::kWebXrGamepadModule.name,
       unified_consent::kUnifiedConsent.name,
-      switches::kSyncUSSBookmarks.name,
```
Comment (bridiver) [inline app/brave_main_delegate.cc]: header in this file as well? https://github.com/brave/brave-core/pull/4324/files#diff-931f170b705cce8579670910b176bdf2L16

Comment (darkdh) [reply app/brave_main_delegate.cc]: it is required for `switches::kSyncServiceURL`

**Comment by btlechowski** [comment]
I found a serious regression in 1.5.x (https://github.com/brave/brave-browser/issues/7893) and therefor we should wait with uplift till the issue is resolved.

**Comment by btlechowski** [comment]
Verification passed on

Brave | 1.5.75 Chromium: 80.0.3987.66 (Official Build) nightly (64-bit)
-- | --
Revision | 56ea77d65c47ecbb1180b99f860d85b51117864a-refs/branch-heads/3987@{#662}
OS | Windows 7 Service Pack 1 (Build 7601.24544)

Brave | 1.5.75 Chromium: 80.0.3987.66 (Official Build) nightly (64-bit)
-- | --
Revision | 56ea77d65c47ecbb1180b99f860d85b51117864a-refs/branch-heads/3987@{#662}
OS | Ubuntu 18.04 LTS

Verified test plan from https://github.com/brave/brave-core/pull/4324#issue-360251085
Also test by switching the devices.
Tested on clean and upgrade profile.

[Pass] Add (1)
[Pass] Add with existing bookmarks (2)
[Pass] Add with same order (3)
[Pass] Move
[Pass] Modify
[Pass] Reorder (1)
[Pass] Reorder (2)
[Pass] Delete
[Pass] Migration

**Pull request merged**